inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/annotation/DynamicWorkflowDocumentNavigationActionBarExtension.java

```diff
@@ -91,53 +115,84 @@ public Panel createActionBarItem(String aId, AnnotationPageBase aPage)
-                //Check if it exist or is NEW
-                if (documentService.listAnnotatableDocuments(project,user).get(doc) == null ||
-                    documentService.listAnnotatableDocuments(project,user).get(doc).
-                        getState().equals(NEW)) {
-                    //This document had the state NEW, load it
-                    aPage.getModelObject().setDocument(doc, documentService.
-                        listSourceDocuments(project));
-
-                    Optional<AjaxRequestTarget> target = RequestCycle.get().
-                        find(AjaxRequestTarget.class);
-                    aPage.actionLoadDocument(target.orElse(null));
-                    return;
-                }
-            }
+
+        // No annotation documents in the state INPROGRESS, now select a new one
+        // depending on the workload strategy selected
+        WorkloadManager currentWorkload = workloadManagementService
+                .loadOrCreateWorkloadManagerConfiguration(project);
+
+        // Get all documents for which the state is NEW, or which have not been created yet.
+        List<SourceDocument> sourceDocuments = workloadManagementService
+                .getAnnotationDocumentListForUser(project, user);
+
+        // Switch for all workflow types which are available. If a new one is created
+        // simply add a new case here.
+        switch (dynamicWorkloadExtension.readTraits(currentWorkload).getWorkflowType()) {
+
+        // Go through all documents in a random order
+        case (RANDOMIZED_WORKFLOW):
+            // Shuffle the List then call loadNewDocument
+            Collections.shuffle(sourceDocuments);
+            loadNewDocument(sourceDocuments, project, currentWorkload, aPage, target);
+            break;
+
+        // Default workflow selected, nothing to change in the list
+        default:
+            loadNewDocument(sourceDocuments, project, currentWorkload, aPage, target);
+            break;
         }
-        //No documents left
-        if (aPage.getModelObject().getDocument() == null) {
-            aPage.setResponsePage(aPage.getApplication().getHomePage());
-            aPage.getSession().info("There are no more documents to annotate available for you. Please contact your project supervisor.");
+    }
+
+    private void loadNewDocument(List<SourceDocument> aSourceDocuments, Project aProject,
+                                 WorkloadManager aCurrentWorkload, AnnotationPageBase aPage,
+                                 Optional<AjaxRequestTarget> aTarget)
+    {
+        // Go through all documents of the list
+        for (SourceDocument doc : aSourceDocuments) {
+            // Check if there are less annotators working on the selected document than
+            // the default number of annotation set by the project manager
+            if ((workloadManagementService.getAmountOfUsersWorkingOnADocument(doc, aProject)
```

Comment:
Looks like duplicated code. Maybe move to a utility class or to a base class for workflow menubar extensions?

Reply:
Actually, the logic that checks the number of people working on a document against the traits should be moved into the `nextDocument` method that I described earlier. The code here should basically be something similar to:

```
aPage.getModelObject().setDocument(workflowExtension.nextDocument(workflowManager, candidates),
                         documentService.listSourceDocuments(aProject));
                 aPage.actionLoadDocument(aTarget.orElse(null));
```